solution/1400-1499/1438.Longest Continuous Subarray With Absolute Diff Less Than or Equal to Limit/README.md

```diff
@@ -1050,136 +1050,167 @@ func (q Deque) Get(i int) int {
 
 ```ts
 function longestSubarray(nums: number[], limit: number): number {
-    const n = nums.length;
-    let [l, r] = [0, n];
-    const check = (k: number): boolean => {
-        const minq = new Deque<number>();
-        const maxq = new Deque<number>();
-        for (let i = 0; i < n; ++i) {
-            while (!minq.isEmpty() && i - minq.frontValue()! + 1 > k) {
-                minq.popFront();
-            }
-            while (!maxq.isEmpty() && i - maxq.frontValue()! + 1 > k) {
-                maxq.popFront();
-            }
-            while (!minq.isEmpty() && nums[minq.backValue()!] >= nums[i]) {
-                minq.popBack();
-            }
-            while (!maxq.isEmpty() && nums[maxq.backValue()!] <= nums[i]) {
-                maxq.popBack();
-            }
-            minq.pushBack(i);
-            maxq.pushBack(i);
-            if (i >= k - 1 && nums[maxq.frontValue()!] - nums[minq.frontValue()!] <= limit) {
-                return true;
-            }
-        }
-        return false;
-    };
-    while (l < r) {
-        const mid = (l + r + 1) >> 1;
-        if (check(mid)) {
-            l = mid;
+    let ans = 0;
+    let l = 0;
+    const maxQueue: number[] = [];
+    const minQueue: number[] = [];
+
+    for (let r = 0; r < nums.length; r++) {
+        while (maxQueue.length && nums[maxQueue.at(-1)!] < nums[r]) maxQueue.pop();
+        while (minQueue.length && nums[minQueue.at(-1)!] > nums[r]) minQueue.pop();
+        maxQueue.push(r);
+        minQueue.push(r);
+
+        const diff = nums[maxQueue[0]] - nums[minQueue[0]];
+
+        if (diff <= limit) {
+            ans = Math.max(ans, r - l + 1);
         } else {
-            r = mid - 1;
+            l++;
+
+            if (maxQueue[0] < l) maxQueue.shift();
```

Comment:
In JavaScript and TypeScript, the time complexity of the `Array.prototype.shift()` method should be \(O(n)\), right?

Reply:
Unfortunately, it does.

Reply:
Yep )
I have seen `shift` repeatedly in some solutions on LeetCode with a best time of execution 
All I can say is the practice is better then the theory )